library/core/src/iter/adapters/skip.rs

```diff
@@ -33,21 +33,29 @@ where
     #[inline]
     fn next(&mut self) -> Option<I::Item> {
         if unlikely(self.n > 0) {
-            self.iter.nth(crate::mem::take(&mut self.n) - 1)?;
+            self.iter.nth(crate::mem::take(&mut self.n))
+        } else {
+            self.iter.next()
         }
-        self.iter.next()
     }
 
     #[inline]
     fn nth(&mut self, n: usize) -> Option<I::Item> {
-        // Can't just add n + self.n due to overflow.
         if self.n > 0 {
-            let to_skip = self.n;
+            // Checked add to handle overflow case.
+            let n = self.n.checked_add(n).unwrap_or_else(|| {
+                // In case of overflow, load skip value, before loading `n`.
+                // Because the amount of elements to iterate is beyond `usize::MAX`, this
+                // is split into two `nth` calls where the `skip` `nth` call is discarded.
+                self.iter.nth(self.n - 1);
```

Comment:
If this returns `None`, we used to early-exit, but now the code will call self.iter.nth(n) and potentially return something on a non-Fused iterator. I think we should preserve that behavior and add a test for this condition.

Reply:
good catch, my mistake for not being thorough! Test added